cmd/dagger/call.go

```diff
@@ -101,9 +113,18 @@ var callCmd = &FuncCommand{
 				writer = io.MultiWriter(writer, file)
 			}
 
+			// especially useful for lists and maps
```

Comment:
I like having `--json` as a separate flag since it makes a difference in the case of e.g. a string return where the difference between `thestring` and `"thestring"` being written to output may actually matter.

But I do wonder if for the list+map case if it would make sense to just use json by default. In the current behavior we print the elements/fields individually as lines, but at that point I feel like we may as well just print them in a json structured format. Either the user actually wanted them as json or even if they didn't care I don't think printing them in that format hurts anything, if that makes sense?

Reply:
What I intended to do with that was put each slice into a progrock group. For maps, you can add the key as label.

Reply:
(didn't finish) But that would only work in the TUI. I've gone back and forth with other instances that I wanted to automatically use JSON, but wasn't sure about how to conciliate with the explicitness of `--json`.

I was on the fence, but if that makes sense to you I also lean more towards automatically setting as JSON, between what we have now and something better later. I have an issue with encoding from MarshalIndent though. It converts "special chars" and makes the text unreadable. JSON is meant to be read by machines.

Reply:
The map should happen when you return a list of objects. If, from those objects, you get multiline output like from stdout, JSON isn't a good fit for human consumption. I do a lot of this with the Python dev module.

Here's a practical example. The `all-tests` function returns two `Test` instances, for both Python versions that we support (3.10 and 3.11). I can sub-select `unit` for just the unit tests, and `stdout` to get their output:

This is now:

```console
❯ dagger -m sdk/python/dev call all-tests unit stdout
============================= test session starts ==============================
platform linux -- Python 3.10.13, pytest-7.4.4, pluggy-1.3.0
rootdir: /work
configfile: pyproject.toml
testpaths: tests/
plugins: anyio-4.2.0, subprocess-1.5.0, mock-3.12.0, lazy-fixture-0.6.3, httpx-0.28.0
collected 220 items / 31 deselected / 189 selected

tests/client/test_codegen.py ........................................... [ 22%]
..................                                                       [ 32%]
tests/client/test_config.py ....                                         [ 34%]
tests/client/test_connection_errors.py ....                              [ 36%]
tests/client/test_execute_errors.py .......                              [ 40%]
tests/client/test_inputs.py .........................                    [ 53%]
tests/client/test_response.py .......                                    [ 57%]
tests/engine/test_cli.py ......................                          [ 68%]
tests/modules/test_forward_reference.py .                                [ 69%]
tests/modules/test_registration.py ..........                            [ 74%]
tests/modules/test_results.py ....................                       [ 85%]
tests/modules/test_utils.py ............................                 [100%]

====================== 189 passed, 31 deselected in 0.85s ======================

============================= test session starts ==============================
platform linux -- Python 3.11.7, pytest-7.4.4, pluggy-1.3.0
rootdir: /work
configfile: pyproject.toml
testpaths: tests/
plugins: anyio-4.2.0, subprocess-1.5.0, mock-3.12.0, lazy-fixture-0.6.3, httpx-0.28.0
collected 220 items / 31 deselected / 189 selected

tests/client/test_codegen.py ........................................... [ 22%]
..................                                                       [ 32%]
tests/client/test_config.py ....                                         [ 34%]
tests/client/test_connection_errors.py ....                              [ 36%]
tests/client/test_execute_errors.py .......                              [ 40%]
tests/client/test_inputs.py .........................                    [ 53%]
tests/client/test_response.py .......                                    [ 57%]
tests/engine/test_cli.py ......................                          [ 68%]
tests/modules/test_forward_reference.py .                                [ 69%]
tests/modules/test_registration.py ..........                            [ 74%]
tests/modules/test_results.py ....................                       [ 85%]
tests/modules/test_utils.py ............................                 [100%]

====================== 189 passed, 31 deselected in 0.77s ======================
```

Well, the output from pytest is useful on its own. I know which is which.

Now with `--json`:

```console
❯ dagger -m sdk/python/dev call all-tests unit stdout
[
    {
        "unit": {
            "stdout": "============================= test session starts ==============================\nplatform linux -- Python 3.10.13, pytest-7.4.4, pluggy-1.3.0\nrootdir: /work\nconfigfile: pyproject.toml\ntestpaths: tests/\nplugins: anyio-4.2.0, subprocess-1.5.0, mock-3.12.0, lazy-fixture-0.6.3, httpx-0.28.0\ncollected 220 items / 31 deselected / 189 selected\n\ntests/client/test_codegen.py ........................................... [ 22%]\n..................                                                       [ 32%]\ntests/client/test_config.py ....                                         [ 34%]\ntests/client/test_connection_errors.py ....                              [ 36%]\ntests/client/test_execute_errors.py .......                              [ 40%]\ntests/client/test_inputs.py .........................                    [ 53%]\ntests/client/test_response.py .......                                    [ 57%]\ntests/engine/test_cli.py ......................                          [ 68%]\ntests/modules/test_forward_reference.py .                                [ 69%]\ntests/modules/test_registration.py ..........                            [ 74%]\ntests/modules/test_results.py ....................                       [ 85%]\ntests/modules/test_utils.py ............................                 [100%]\n\n====================== 189 passed, 31 deselected in 0.85s ======================\n"
        }
    },
    {
        "unit": {
            "stdout": "============================= test session starts ==============================\nplatform linux -- Python 3.11.7, pytest-7.4.4, pluggy-1.3.0\nrootdir: /work\nconfigfile: pyproject.toml\ntestpaths: tests/\nplugins: anyio-4.2.0, subprocess-1.5.0, mock-3.12.0, lazy-fixture-0.6.3, httpx-0.28.0\ncollected 220 items / 31 deselected / 189 selected\n\ntests/client/test_codegen.py ........................................... [ 22%]\n..................                                                       [ 32%]\ntests/client/test_config.py ....                                         [ 34%]\ntests/client/test_connection_errors.py ....                              [ 36%]\ntests/client/test_execute_errors.py .......                              [ 40%]\ntests/client/test_inputs.py .........................                    [ 53%]\ntests/client/test_response.py .......                                    [ 57%]\ntests/engine/test_cli.py ......................                          [ 68%]\ntests/modules/test_forward_reference.py .                                [ 69%]\ntests/modules/test_registration.py ..........                            [ 74%]\ntests/modules/test_results.py ....................                       [ 85%]\ntests/modules/test_utils.py ............................                 [100%]\n\n====================== 189 passed, 31 deselected in 0.77s ======================\n"
        }
    }
]
```

---


Here's a simpler example. I can see the SDK's pinned dependencies:

```console
❯ dagger -m sdk/python/dev call deps requirements contents
anyio>=3.6.2
beartype>=0.11.0
cattrs>=22.2.0
gql>=3.4.0
graphql-core>=3.2.3
httpx>=0.23.1
platformdirs>=2.6.2
rich>=10.11.0
typing-extensions>=4.8.0
```

But also the dependencies for all dev "environments" (tests, docs, lint):

```console
❯ dagger -m sdk/python/dev call all-deps requirements contents
anyio>=3.6.2
beartype>=0.11.0
cattrs>=22.2.0
gql>=3.4.0
graphql-core>=3.2.3
httpx>=0.23.1
platformdirs>=2.6.2
rich>=10.11.0
typing-extensions>=4.8.0

aiohttp>=3.8.6
pytest-httpx>=0.21.3
pytest-lazy-fixture>=0.6.3
pytest-mock>=3.10.0
pytest-subprocess>=1.4.2
pytest>=7.2.0

sphinx-rtd-theme>=1.3.0
sphinx>=5.3.0

black>=22.3.0
mypy>=0.942
ruff>=0.0.218
```

Grrr... I wish I could see which file each of these contents are from! 😮‍💨

Let's try with JSON:

```console
❯ dagger -m sdk/python/dev call all-deps requirements contents --json
[
    {
        "requirements": {
            "contents": "anyio\u003e=3.6.2\nbeartype\u003e=0.11.0\ncattrs\u003e=22.2.0\ngql\u003e=3.4.0\ngraphql-core\u003e=3.2.3\nhttpx\u003e=0.23.1\nplatformdirs\u003e=2.6.2\nrich\u003e=10.11.0\ntyping-extensions\u003e=4.8.0\n"
        }
    },
    {
        "requirements": {
            "contents": "aiohttp\u003e=3.8.6\npytest-httpx\u003e=0.21.3\npytest-lazy-fixture\u003e=0.6.3\npytest-mock\u003e=3.10.0\npytest-subprocess\u003e=1.4.2\npytest\u003e=7.2.0\n"
        }
    },
    {
        "requirements": {
            "contents": "sphinx-rtd-theme\u003e=1.3.0\nsphinx\u003e=5.3.0\n"
        }
    },
    {
        "requirements": {
            "contents": "black\u003e=22.3.0\nmypy\u003e=0.942\nruff\u003e=0.0.218\n"
        }
    }
]
```

Uh, I wish I had the file's name besides contents. Would need _simple objects_ for that.

Reply:
> Here's a practical example

All great points! Makes sense to me to keep it as is but support `--json` as an option 👍

> It converts "special chars"

On this specifically, Go's default json marshal escapes HTML-related characters, I *think* that's what's happening there with `>` becoming `\u003e`. If so, it would probably be good to update `--json` to make an explicit `json.NewEncoder` and then set https://pkg.go.dev/encoding/json#Encoder.SetEscapeHTML to `false`.

Reply:
Thanks for the tip, I made this change.